src/cross_sections.cpp

```diff
@@ -136,26 +145,229 @@ void read_cross_sections_xml()
+            LibraryKey key {Library::Type::photon, element};
+            int idx = data::library_map[key];
+            std::string& filename = data::libraries[idx].path_;
+            int i_element = data::element_map[element];
+            write_message("Reading " + element + " from " + filename, 6);
+
+            // Open file and make sure version is sufficient
+            hid_t file_id = file_open(filename, 'r');
+            check_data_version(file_id);
+
+            // Read element data from HDF5
+            hid_t group = open_group(file_id, element.c_str());
+            data::elements.emplace_back(group, data::elements.size());
+
+            // Determine if minimum/maximum energy for this element is greater/less than
+            // the previous
+            const auto& elem {data::elements.back()};
+            if (elem.energy_.size() >= 1) {
+              // TODO: off-by-one
+              int photon = static_cast<int>(ParticleType::photon) - 1;
+              int n = elem.energy_.size();
+              data::energy_min[photon] = std::max(data::energy_min[photon],
+                std::exp(elem.energy_(1)));
+              data::energy_max[photon] = std::min(data::energy_max[photon],
+                std::exp(elem.energy_(n - 1)));
+            }
+
+            close_group(group);
+            file_close(file_id);
+
+            // Add element to set
+            already_read.insert(element);
+          }
+        }
+
+        // Read multipole file into the appropriate entry on the nuclides array
+        if (settings::temperature_multipole) read_multipole_data(i_nuclide);
+      }
+    }
+  }
+
+  for (auto& mat : model::materials) {
+    for (const auto& table : mat->thermal_tables_) {
+      // Get name of S(a,b) table
+      int i_table = table.index_table;
+      std::string& name = thermal_names[i_table];
+
+      if (already_read.find(name) == already_read.end()) {
+        LibraryKey key {Library::Type::thermal, name};
+        int idx = data::library_map[key];
+        std::string& filename = data::libraries[idx].path_;
+
+        write_message("Reading " + name + " from " + filename, 6);
+
+        // Open file and make sure version matches
+        hid_t file_id = file_open(filename, 'r');
+        check_data_version(file_id);
+
+        // Read thermal scattering data from HDF5
+        hid_t group = open_group(file_id, name.c_str());
+        data::thermal_scatt.push_back(std::make_unique<ThermalScattering>(
+          group, thermal_temps[i_table]));
+        close_group(group);
+        file_close(file_id);
+
+        // Add name to dictionary
+        already_read.insert(name);
+      }
+    } // thermal_tables_
+
+    // Finish setting up materials (normalizing densities, etc.)
+    mat->finalize();
+  } // materials
+
+
+  // Set up logarithmic grid for nuclides
+  for (auto& nuc : data::nuclides) {
+    nuc->init_grid();
+  }
+  int neutron = static_cast<int>(ParticleType::neutron) - 1;
```

Comment:
TODO: off-by-one

Reply:
Also, what remains for this off-by-one to be removed? I think there are a few instances of `Particle % type` being called from tracking and tallies but thats it, right?  

I'm only partially through this PR, but if there are many more of these off-by-ones introduced because of only a few lines in Fortran, maybe it'd be better to fix the Fortran side instead

Reply:
It would probably be a pretty easy change actually; I'll look into it.